### PR TITLE
Enhance osdmon check failure

### PR DESCRIPTION
**Pull request by xiaoxichen**
In many case OSDs in an isolated(public/cluster connection lost but osd<->mon is good) node
will report other OSD down to monitor,which usually wrongly mark someone down.

Nowaday the "osd_min_down_reporters", we would like to extend the semantic to
allow it counted by host or rack, thus user could require failure reports from at
least two nodes to mark an OSD down, which shoudl prevent an isoloated host make trouble
to the cluster.

### CODE REVIEW

**Comment by yuyuyu101** [comment]
Actually I think this is a good idea. But it looks a little inflexible, maybe we can make this a config item in crush? Each pool could own its failure level, or this could be associated to the type of "chooseleaf_firstn". What's your options? @liewegas @gregsfortytwo

**Comment by liewegas** [comment]
I also think this is a great idea.  It's inflexibility is similar to

OPTION(mon_osd_down_out_subtree_limit, OPT_STR, "rack")   // smallest crush unit/type that we will not automatically mark out

which I think is fine for now.  That only really breaks down if you have different parts of the same crush hierarchy that have different failure domains... which to me says that the crush types aren't being used properly.  Putting it in crush rules or pool properties doesn't make sense because we're looking at the specific osds at his level.

File: src/common/config_opts.h [hunk 1]
```diff
@@ -274,7 +274,8 @@ OPTION(mon_sync_debug_leader, OPT_INT, -1) // monitor to be used as the sync lea
 OPTION(mon_sync_debug_provider, OPT_INT, -1) // monitor to be used as the sync provider
 OPTION(mon_sync_debug_provider_fallback, OPT_INT, -1) // monitor to be used as fallback if sync provider fails
 OPTION(mon_inject_sync_get_chunk_delay, OPT_DOUBLE, 0)  // inject N second delay on each get_chunk request
-OPTION(mon_osd_min_down_reporters, OPT_INT, 1)   // number of OSDs who need to report a down OSD for it to count
+OPTION(mon_osd_reporter_authorized_level, OPT_STR, "host")   // in what level we will aggregate the reporters
```
Comment (liewegas) [inline src/common/config_opts.h]: mon_osd_reporter_subtree_level ?

**Comment by gregsfortytwo** [comment]
I haven't looked at the code but I like the idea, yes. It's possible it should just subsume the min down reporters entirely as that's what it was largely used for?

**Comment by liewegas** [comment]
I think we could keep the min_down_reporters, but instead make it a min_subtrees type option.. so that if you set the level to host you're basically requiring that (osds from) at least N other hosts report failures.

@xiaoxichen can you please rebase this on wip-bigbang?  It simplifies this code a bit already and should go into the tree first.

Thanks!

**Comment by xiaoxichen** [comment]
@liewegas , rebased on PR #6596

**Comment by xiaoxichen** [comment]
closing this one

**Pull request closed without merging**